File: python/local_context_engine/engine.py
"""The Local Context Engine orchestrator.

This is the main entry point used before calling any LLM API:

    engine = LocalContextEngine()
    engine.add_file("notes.md")
    engine.add_history_jsonl("history.jsonl")
    window = engine.build_context(query="...", max_tokens=4000)
    messages = window.as_messages()
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Dict, Iterable, List, Optional, Tuple

from .cache import TTLCache
from .compression import CompressionResult, build_omitted_note, compress_items
from .index import LocalIndex
from .ingest import (
    chunk_text,
    ingest_directory,
    ingest_file,
    ingest_history_jsonl,
    ingest_text,
    ingest_trajectory_jsonl,
)
from .models import (
    Chunk,
    ContextSection,
    ContextWindow,
    RetrievedItem,
    SourceKind,
    estimate_tokens,
)
from .prioritization import PriorityConfig, prioritize
from .retrieval import RetrievalConfig, retrieve
from .sqlite_index import SqliteIndex


class LocalContextEngine:
    """Provider-agnostic dynamic context window builder."""

    def __init__(
        self,
        index: Optional[LocalIndex | SqliteIndex] = None,
        storage_path: Optional[str | Path] = None,
        backend: str = "auto",
        cache_size: int = 128,
        cache_ttl: float = 60.0,
    ) -> None:
        self.storage_path = Path(storage_path) if storage_path else None
        self.retrieval_cache = TTLCache(maxsize=cache_size, ttl_seconds=cache_ttl)
        self.compression_cache = TTLCache(maxsize=cache_size, ttl_seconds=cache_ttl)
        if index is not None:
            self.index = index
        elif self.storage_path is not None:
            self.index = self._open_index(self.storage_path, backend)
        elif backend == "sqlite":
            self.index = SqliteIndex(":memory:")
        else:
            self.index = LocalIndex()

    @staticmethod
    def _open_index(path: Path, backend: str) -> LocalIndex | SqliteIndex:
        if backend == "sqlite":
            return SqliteIndex(path)
        if backend == "json":
            if path.exists():
                return LocalIndex.load(path)
            return LocalIndex()
        # auto
        if path.exists():
            if path.suffix.lower() in {".db", ".sqlite", ".sqlite3"} or _looks_like_sqlite(path):
                return SqliteIndex(path)
            return LocalIndex.load(path)
        if path.suffix.lower() in {".db", ".sqlite", ".sqlite3"}:
            return SqliteIndex(path)
        return LocalIndex()

    # ------------------------------------------------------------------
    # Ingest API
    # ------------------------------------------------------------------
    def add_chunk(self, chunk: Chunk) -> None:
        self.index.add(chunk)

    def add_chunks(self, chunks: Iterable[Chunk]) -> int:
        return self.index.add_many(chunks)

    def add_text(
        self,
        content: str,
        source: str,
        kind: SourceKind,
        chunk_size: int = 800,
        overlap: int = 80,
        metadata: Optional[Dict] = None,
        timestamp: Optional[float] = None,
        priority: float = 0.0,
    ) -> int:
        chunks = chunk_text(
            content,
            source,
            kind,
            chunk_size=chunk_size,
            overlap=overlap,
            metadata=metadata,
            timestamp=timestamp,
            priority=priority,
        )
        return self.add_chunks(chunks)

    def add_file(self, path: str | Path, kind: Optional[SourceKind] = None) -> int:
        return self.add_chunks(ingest_file(path, kind=kind))

    def add_directory(
        self,
        directory: str | Path,
        kind: Optional[SourceKind] = None,
        extensions: Optional[Iterable[str]] = None,
        ignore: Optional[Iterable[str]] = None,
    ) -> int:
        return self.add_chunks(ingest_directory(directory, kind=kind, extensions=extensions, ignore=ignore))

    def add_history_jsonl(self, path: str | Path, source_name: Optional[str] = None) -> int:
        return self.add_chunks(ingest_history_jsonl(path, source_name=source_name))

    def add_trajectory_jsonl(self, path: str | Path, source_name: Optional[str] = None) -> int:
        return self.add_chunks(ingest_trajectory_jsonl(path, source_name=source_name))

    def add_file_incremental(
        self,
        path: str | Path,
        kind: Optional[SourceKind] = None,
        state_path: Optional[str | Path] = None,
    ) -> Tuple[int, int, int]:
        """Ingest a file only when its mtime/size changed.

        Returns ``(added_count, changed_count, skipped_count)``.
        """
        path = Path(path)
        state_path = Path(state_path) if state_path else self._default_state_path()
        state = self._load_state(state_path)
        stat = path.stat()
        key = str(path.resolve())
        previous = state.get(key)
        if previous is not None and previous.get("mtime") == stat.st_mtime and previous.get("size") == stat.st_size:
            return 0, 0, 1

        count = self.add_file(path, kind=kind)
        state[key] = {"mtime": stat.st_mtime, "size": stat.st_size}
        self._save_state(state_path, state)
        return count, 1 if count else 0, 0

    def add_directory_incremental(
        self,
        directory: str | Path,
        kind: Optional[SourceKind] = None,
        extensions: Optional[Iterable[str]] = None,
        ignore: Optional[Iterable[str]] = None,
        state_path: Optional[str | Path] = None,
    ) -> Tuple[int, int, int]:
        """Ingest only new/changed files in a directory.

        Returns ``(added_count, changed_count, skipped_count)``.
        """
        directory = Path(directory)
        state_path = Path(state_path) if state_path else self._default_state_path()
        state = self._load_state(state_path)
        allowed = set(extensions or [".py", ".ts", ".tsx", ".js", ".jsx", ".md", ".txt", ".json", ".jsonl", ".csv", ".html", ".css"])
        ignored = set(ignore or [])
        current_keys = set()
        total_added = 0
        total_changed = 0
        total_skipped = 0

        for path in sorted(directory.rglob("*")):
            if not path.is_file():
                continue
            if any(part in ignored for part in path.parts):
                continue
            if allowed and path.suffix.lower() not in allowed:
                continue
            key = str(path.resolve())
            current_keys.add(key)
            stat = path.stat()
            previous = state.get(key)
            if previous is not None and previous.get("mtime") == stat.st_mtime and previous.get("size") == stat.st_size:
                total_skipped += 1
                continue
            try:
                count = self.add_file(path, kind=kind)
            except Exception as exc:
                print(f"[ingest] skipped {path}: {exc}")
                continue
            state[key] = {"mtime": stat.st_mtime, "size": stat.st_size}
            total_added += count
            total_changed += 1 if count else 0

        # Drop state entries for files that no longer exist (optional cleanup).
        for key in list(state.keys()):
            if key not in current_keys and Path(key).exists() is False:
                del state[key]

        self._save_state(state_path, state)
        return total_added, total_changed, total_skipped

    def clear_caches(self) -> None:
        self.retrieval_cache.clear()
        self.compression_cache.clear()

    def _default_state_path(self) -> Path:
        if self.storage_path is not None:
            return self.storage_path.with_name(self.storage_path.name + ".state.json")
        return Path(".lce_incremental_state.json")

    @staticmethod
    def _load_state(path: Path) -> Dict:
        if not path.exists():
            return {}
        try:
            return json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            return {}

    @staticmethod
    def _save_state(path: Path, state: Dict) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps(state, ensure_ascii=False, indent=2), encoding="utf-8")

    # ------------------------------------------------------------------
    # Persistence
    # ------------------------------------------------------------------
    def save(self, path: Optional[str | Path] = None) -> None:
        target = Path(path) if path else self.storage_path
        if target is None:
            raise ValueError("No storage path configured; pass one to save().")
        self.index.save(target)

    def load(self, path: Optional[str | Path] = None, backend: str = "auto") -> None:
        target = Path(path) if path else self.storage_path
        if target is None:
            raise ValueError("No storage path configured; pass one to load().")
        self.index = self._open_index(target, backend)
        self.storage_path = target

    # ------------------------------------------------------------------
    # Context building
    # ------------------------------------------------------------------
    def build_context(
        self,
        query: str,
        max_tokens: int = 8000,
        reserved_output_tokens: int = 1024,
        retrieval_config: Optional[RetrievalConfig] = None,
        priority_config: Optional[PriorityConfig] = None,
        include_preferences: bool = True,
    ) -> ContextWindow:
        """Assemble a dynamic context window for a query.

        The engine always reserves room for the model's output, then fills the
        remaining budget with preferences first, followed by retrieved history,
        trajectory, code and files. Lower-priority items are compressed or
        omitted with a visible note.
        """
        budget = max(0, max_tokens - reserved_output_tokens)
        if budget <= 0:
            raise ValueError("max_tokens must be greater than reserved_output_tokens")

        # 1. Retrieve and prioritize (with retrieval cache).
        retrieval_key = self._retrieval_cache_key(query, retrieval_config)
        retrieved = self.retrieval_cache.get(retrieval_key)
        if retrieved is None:
            retrieved = retrieve(self.index, query, retrieval_config)
            self.retrieval_cache.set(retrieval_key, retrieved)
        prioritized = prioritize(retrieved, priority_config)

        # 2. Always-available user preferences section.
        pref_section = None
        omitted_count = 0
        omitted_tokens = 0
        if include_preferences:
            pref_chunks = self.index.chunks_by_kind(SourceKind.PREFERENCE)
            if pref_chunks:
                pref_items = [
                    RetrievedItem(chunk=c, score=10.0 - i * 0.001, reason="always-include preference")
                    for i, c in enumerate(pref_chunks)
                ]
                pref_key = self._compression_cache_key(query, budget, pref_items)
                pref_result = self.compression_cache.get(pref_key)
                if pref_result is None:
                    pref_result = compress_items(pref_items, budget, query=query)
                    self.compression_cache.set(pref_key, pref_result)
                pref_section = self._section_from_result(
                    "User Preferences", pref_result, kind=SourceKind.PREFERENCE
                )
                omitted_count += pref_result.omitted_count
                omitted_tokens += pref_result.omitted_tokens

        used = pref_section.tokens if pref_section else 0
        remaining = max(0, budget - used)

        # 3. Compress the remaining retrieved context into the leftover budget.
        non_pref_items = [item for item in prioritized if item.chunk.kind != SourceKind.PREFERENCE]
        non_pref_key = self._compression_cache_key(query, remaining, non_pref_items)
        non_pref_result = self.compression_cache.get(non_pref_key)
        if non_pref_result is None:
            non_pref_result = compress_items(non_pref_items, remaining, query=query)
            self.compression_cache.set(non_pref_key, non_pref_result)
        omitted_count += non_pref_result.omitted_count
        omitted_tokens += non_pref_result.omitted_tokens

        sections: List[ContextSection] = []
        if pref_section:
            sections.append(pref_section)

        # 4. Group packed items into named sections.
        sections.extend(self._group_into_sections(non_pref_result))

        # 5. Add an omitted-memory note if there is room.
        note = build_omitted_note(omitted_count, omitted_tokens)
        if note:
            note_tokens = estimate_tokens(note)
            used_after_sections = sum(s.tokens for s in sections)
            if used_after_sections + note_tokens <= budget:
                sections.append(
                    ContextSection(
                        title="Omitted Memory",
                        content=note,
                        tokens=note_tokens,
                        compressed=True,
                    )
                )

        total_tokens = sum(s.tokens for s in sections)
        return ContextWindow(
            query=query,
            sections=sections,
            max_tokens=max_tokens,
            reserved_output_tokens=reserved_output_tokens,
            token_budget=budget,
            total_tokens=total_tokens,
            omitted_count=omitted_count,
            omitted_tokens=omitted_tokens,
        )

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------
    def _retrieval_cache_key(self, query: str, config: Optional[RetrievalConfig]) -> tuple:
        return (
            "retrieve",
            self.index.revision,
            query.strip().lower(),
            self._retrieval_config_fingerprint(config),
        )

    @staticmethod
    def _retrieval_config_fingerprint(config: Optional[RetrievalConfig]) -> Optional[tuple]:
        if config is None:
            return None
        return (
            config.relevance_weight,
            config.recency_weight,
            config.source_weight,
            config.priority_weight,
            config.recency_half_life_days,
            config.top_k,
            tuple(sorted((k.value, v) for k, v in config.source_weights.items())),
        )

    def _compression_cache_key(self, query: str, budget: int, items: List[RetrievedItem]) -> tuple:
        # Include index revision so a changed index cannot return stale results.
        item_sig = tuple((item.chunk.id, round(item.score, 6)) for item in items)
        return ("compress", self.index.revision, query.strip().lower(), budget, item_sig)

    @staticmethod
    def _section_from_result(
        title: str,
        result: CompressionResult,
        kind: SourceKind,
    ) -> ContextSection:
        packed = result.packed
        content = LocalContextEngine._render_packed(packed)
        tokens = sum(p.tokens for p in packed)
        return ContextSection(
            title=title,
            content=content,
            tokens=tokens,
            items=[p.item for p in packed],
            compressed=any(p.compressed for p in packed),
        )

    @staticmethod
    def _group_into_sections(result: CompressionResult) -> List[ContextSection]:
        """Group packed items by SourceKind into stable, human-readable sections."""
        titles = {
            SourceKind.HISTORY: "Relevant History",
            SourceKind.TRAJECTORY: "Agent Trajectory",
            SourceKind.CODE: "Relevant Code",
            SourceKind.FILE: "Relevant Files",
        }
        grouped: Dict[SourceKind, List] = {kind: [] for kind in titles}
        for packed in result.packed:
            kind = packed.item.chunk.kind
            if kind in grouped:
                grouped[kind].append(packed)

        sections: List[ContextSection] = []
        for kind in titles:
            packed = grouped[kind]
            if not packed:
                continue
            content = LocalContextEngine._render_packed(packed)
            tokens = sum(p.tokens for p in packed)
            sections.append(
                ContextSection(
                    title=titles[kind],
                    content=content,
                    tokens=tokens,
                    items=[p.item for p in packed],
                    compressed=any(p.compressed for p in packed),
                )
            )
        return sections

    @staticmethod
    def _render_packed(packed) -> str:
        parts: List[str] = []
        for p in packed:
            chunk = p.item.chunk
            source_label = chunk.source
            if chunk.metadata.get("path"):
                source_label = chunk.metadata["path"]
            parts.append(f"[{chunk.kind.value} | {source_label}]\n{p.content}")
        return "\n\n".join(parts)


def _looks_like_sqlite(path: Path) -> bool:
    """Detect a SQLite database by its 16-byte file header."""
    try:
        with path.open("rb") as fh:
            return fh.read(16) == b"SQLite format 3\x00"
    except OSError:
        return False

File: engine/engine.js
// The Local Context Engine orchestrator (JS port).

import { closeSync, existsSync, mkdirSync, openSync, readFileSync, readdirSync, readSync, statSync, writeFileSync } from "node:fs";
import { dirname, extname, join, resolve } from "node:path";

import { TTLCache } from "./cache.js";
import { buildOmittedNote, compressItems } from "./compression.js";
import { LocalIndex } from "./index.js";
import { SqliteIndex } from "./sqlite_index.js";
import {
  chunkText,
  ingestDirectory,
  ingestFile,
  ingestHistoryJsonl,
  ingestTrajectoryJsonl,
} from "./ingest.js";
import { Chunk, ContextSection, ContextWindow, RetrievedItem, SourceKind, estimateTokens } from "./models.js";
import { PriorityConfig, prioritize } from "./prioritization.js";
import { RetrievalConfig, retrieve } from "./retrieval.js";

const DEFAULT_DIRECTORY_EXTENSIONS = [".py", ".ts", ".tsx", ".js", ".jsx", ".md", ".txt", ".json", ".jsonl", ".csv", ".html", ".css"];

export class LocalContextEngine {
  constructor({ index = null, storagePath = null, backend = "auto", cacheSize = 128, cacheTtl = 60 } = {}) {
    this.storagePath = storagePath ? String(storagePath) : null;
    this.retrievalCache = new TTLCache({ maxsize: cacheSize, ttlSeconds: cacheTtl });
    this.compressionCache = new TTLCache({ maxsize: cacheSize, ttlSeconds: cacheTtl });
    if (index) {
      this.index = index;
    } else if (this.storagePath) {
      this.index = this._openIndex(this.storagePath, backend);
    } else if (backend === "sqlite") {
      this.index = new SqliteIndex(":memory:");
    } else {
      this.index = new LocalIndex();
    }
  }

  // ------------------------------------------------------------------
  // Ingest API
  // ------------------------------------------------------------------
  addChunk(chunk) {
    this.index.add(chunk);
  }

  addChunks(chunks) {
    return this.index.addMany(chunks);
  }

  addText({ content, source, kind, chunkSize = 800, overlap = 80, metadata = {}, timestamp = null, priority = 0 }) {
    const chunks = chunkText({ content, source, kind, chunkSize, overlap, metadata, timestamp, priority });
    return this.addChunks(chunks);
  }

  addFile(filePath, kind = null) {
    return this.addChunks(ingestFile(filePath, kind));
  }

  addDirectory(directory, { kind = null, extensions = null, ignore = [] } = {}) {
    return this.addChunks(ingestDirectory(directory, { kind, extensions, ignore }));
  }

  addHistoryJsonl(filePath, sourceName = null) {
    return this.addChunks(ingestHistoryJsonl(filePath, sourceName));
  }

  addTrajectoryJsonl(filePath, sourceName = null) {
    return this.addChunks(ingestTrajectoryJsonl(filePath, sourceName));
  }

  addFileIncremental({ path, kind = null, statePath = null } = {}) {
    const stateFile = statePath ? String(statePath) : this._defaultStatePath();
    const state = this._loadState(stateFile);
    const key = resolve(path);
    const stat = statSync(path);
    const previous = state[key];
    if (previous && previous.mtime === stat.mtimeMs && previous.size === stat.size) {
      return { added: 0, changed: 0, skipped: 1 };
    }
    const count = this.addFile(path, kind);
    state[key] = { mtime: stat.mtimeMs, size: stat.size };
    this._saveState(stateFile, state);
    return { added: count, changed: count > 0 ? 1 : 0, skipped: 0 };
  }

  addDirectoryIncremental({ directory, kind = null, extensions = null, ignore = [], statePath = null } = {}) {
    const stateFile = statePath ? String(statePath) : this._defaultStatePath();
    const state = this._loadState(stateFile);
    const allowed = new Set(extensions || DEFAULT_DIRECTORY_EXTENSIONS);
    const ignored = new Set(ignore || []);
    let totalAdded = 0;
    let totalChanged = 0;
    let totalSkipped = 0;
    const currentKeys = new Set();

    const walk = (dir) => {
      const entries = readdirSync(dir, { withFileTypes: true });
      for (const entry of entries) {
        const full = join(dir, entry.name);
        if (entry.isDirectory()) {
          walk(full);
          continue;
        }
        if (!entry.isFile()) continue;
        if (ignored.size > 0 && full.split(/[\\/]/).some((part) => ignored.has(part))) continue;
        if (allowed.size > 0 && !allowed.has(extname(full).toLowerCase())) continue;
        const key = resolve(full);
        currentKeys.add(key);
        const stat = statSync(full);
        const previous = state[key];
        if (previous && previous.mtime === stat.mtimeMs && previous.size === stat.size) {
          totalSkipped += 1;
          continue;
        }
        try {
          const count = this.addFile(full, kind);
          state[key] = { mtime: stat.mtimeMs, size: stat.size };
          totalAdded += count;
          totalChanged += count > 0 ? 1 : 0;
        } catch {
          // skip unreadable file
        }
      }
    };
    walk(directory);

    for (const key of Object.keys(state)) {
      if (!currentKeys.has(key) && !existsSync(key)) delete state[key];
    }
    this._saveState(stateFile, state);
    return { added: totalAdded, changed: totalChanged, skipped: totalSkipped };
  }

  clearCaches() {
    this.retrievalCache.clear();
    this.compressionCache.clear();
  }

  // ------------------------------------------------------------------
  // Persistence
  // ------------------------------------------------------------------
  save(path = null) {
    const target = path ? String(path) : this.storagePath;
    if (!target) throw new Error("No storage path configured; pass one to save().");
    this.index.save(target);
  }

  load(path = null, backend = "auto") {
    const target = path ? String(path) : this.storagePath;
    if (!target) throw new Error("No storage path configured; pass one to load().");
    this.index = this._openIndex(target, backend);
    this.storagePath = target;
  }

  // ------------------------------------------------------------------
  // Context building
  // ------------------------------------------------------------------
  buildContext({ query, maxTokens = 8000, reservedOutputTokens = 1024, retrievalConfig = null, priorityConfig = null, includePreferences = true } = {}) {
    const budget = Math.max(0, maxTokens - reservedOutputTokens);
    if (budget <= 0) throw new Error("maxTokens must be greater than reservedOutputTokens");

    const retrievalKey = this._retrievalCacheKey(query, retrievalConfig);
    let retrieved = this.retrievalCache.get(retrievalKey);
    if (!retrieved) {
      retrieved = retrieve(this.index, query, retrievalConfig);
      this.retrievalCache.set(retrievalKey, retrieved);
    }
    const prioritized = prioritize(retrieved, priorityConfig);

    let prefSection = null;
    let omittedCount = 0;
    let omittedTokens = 0;
    if (includePreferences) {
      const prefChunks = this.index.chunksByKind(SourceKind.PREFERENCE);
      if (prefChunks.length > 0) {
        const prefItems = prefChunks.map((chunk, i) => new RetrievedItem({ chunk, score: 10.0 - i * 0.001, reason: "always-include preference" }));
        const prefKey = this._compressionCacheKey(query, budget, prefItems);
        let prefResult = this.compressionCache.get(prefKey);
        if (!prefResult) {
          prefResult = compressItems(prefItems, budget, query);
          this.compressionCache.set(prefKey, prefResult);
        }
        prefSection = this._sectionFromResult("User Preferences", prefResult);
        omittedCount += prefResult.omittedCount;
        omittedTokens += prefResult.omittedTokens;
      }
    }

    const used = prefSection ? prefSection.tokens : 0;
    const remaining = Math.max(0, budget - used);
    const nonPrefItems = prioritized.filter((item) => item.chunk.kind !== SourceKind.PREFERENCE);
    const nonPrefKey = this._compressionCacheKey(query, remaining, nonPrefItems);
    let nonPrefResult = this.compressionCache.get(nonPrefKey);
    if (!nonPrefResult) {
      nonPrefResult = compressItems(nonPrefItems, remaining, query);
      this.compressionCache.set(nonPrefKey, nonPrefResult);
    }
    omittedCount += nonPrefResult.omittedCount;
    omittedTokens += nonPrefResult.omittedTokens;

    const sections = [];
    if (prefSection) sections.push(prefSection);
    sections.push(...this._groupIntoSections(nonPrefResult));

    const note = buildOmittedNote(omittedCount, omittedTokens);
    if (note) {
      const noteTokens = estimateTokens(note);
      const usedAfterSections = sections.reduce((sum, section) => sum + section.tokens, 0);
      if (usedAfterSections + noteTokens <= budget) {
        sections.push(new ContextSection({ title: "Omitted Memory", content: note, tokens: noteTokens, compressed: true }));
      }
    }

    const totalTokens = sections.reduce((sum, section) => sum + section.tokens, 0);
    return new ContextWindow({
      query,
      sections,
      maxTokens,
      reservedOutputTokens,
      tokenBudget: budget,
      totalTokens,
      omittedCount,
      omittedTokens,
    });
  }

  // ------------------------------------------------------------------
  // Internal helpers
  // ------------------------------------------------------------------
  _retrievalCacheKey(query, config) {
    return ["retrieve", this.index.revision, String(query).trim().toLowerCase(), this._retrievalConfigFingerprint(config)];
  }

  _retrievalConfigFingerprint(config) {
    if (!config) return null;
    return [
      config.relevanceWeight,
      config.recencyWeight,
      config.sourceWeight,
      config.priorityWeight,
      config.recencyHalfLifeDays,
      config.topK,
      Object.entries(config.sourceWeights).sort(([a], [b]) => a.localeCompare(b)),
    ];
  }

  _compressionCacheKey(query, budget, items) {
    const itemSig = items.map((item) => [item.chunk.id, Math.round(item.score * 1e6) / 1e6]);
    return ["compress", this.index.revision, String(query).trim().toLowerCase(), budget, itemSig];
  }

  _sectionFromResult(title, result) {
    const content = this._renderPacked(result.packed);
    const tokens = result.packed.reduce((sum, packed) => sum + packed.tokens, 0);
    return new ContextSection({
      title,
      content,
      tokens,
      items: result.packed.map((packed) => packed.item),
      compressed: result.packed.some((packed) => packed.compressed),
    });
  }

  _groupIntoSections(result) {
    const titles = {
      [SourceKind.HISTORY]: "Relevant History",
      [SourceKind.TRAJECTORY]: "Agent Trajectory",
      [SourceKind.CODE]: "Relevant Code",
      [SourceKind.FILE]: "Relevant Files",
    };
    const grouped = new Map(Object.keys(titles).map((kind) => [kind, []]));
    for (const packed of result.packed) {
      const kind = packed.item.chunk.kind;
      if (grouped.has(kind)) grouped.get(kind).push(packed);
    }
    const sections = [];
    for (const [kind, packedList] of grouped) {
      if (packedList.length === 0) continue;
      const content = this._renderPacked(packedList);
      const tokens = packedList.reduce((sum, packed) => sum + packed.tokens, 0);
      sections.push(
        new ContextSection({
          title: titles[kind],
          content,
          tokens,
          items: packedList.map((packed) => packed.item),
          compressed: packedList.some((packed) => packed.compressed),
        }),
      );
    }
    return sections;
  }

  _renderPacked(packedList) {
    return packedList
      .map((packed) => {
        const chunk = packed.item.chunk;
        const sourceLabel = chunk.metadata?.path || chunk.source;
        return `[${chunk.kind} | ${sourceLabel}]\n${packed.content}`;
      })
      .join("\n\n");
  }

  _openIndex(path, backend) {
    if (backend === "sqlite") return new SqliteIndex(path);
    if (backend === "json") return existsSync(path) ? LocalIndex.load(path) : new LocalIndex();
    if (existsSync(path)) {
      if (/\.(db|sqlite|sqlite3)$/i.test(path) || looksLikeSqlite(path)) return new SqliteIndex(path);
      return LocalIndex.load(path);
    }
    if (/\.(db|sqlite|sqlite3)$/i.test(path)) return new SqliteIndex(path);
    return new LocalIndex();
  }

  _defaultStatePath() {
    if (this.storagePath) return `${this.storagePath}.state.json`;
    return ".lce_incremental_state.json";
  }

  _loadState(path) {
    if (!existsSync(path)) return {};
    try {
      return JSON.parse(readFileSync(path, "utf8"));
    } catch {
      return {};
    }
  }

  _saveState(path, state) {
    mkdirSync(dirname(path), { recursive: true });
    writeFileSync(path, JSON.stringify(state, null, 2), "utf8");
  }
}

function looksLikeSqlite(path) {
  try {
    const fd = openSync(path, "r");
    try {
      const buffer = Buffer.alloc(16);
      readSync(fd, buffer, 0, 16, 0);
      return buffer.toString("latin1") === "SQLite format 3\u0000";
    } finally {
      closeSync(fd);
    }
  } catch {
    return false;
  }
}
